src/main/scala/mesosphere/marathon/event/http/HttpEventStreamActor.scala

```diff
@@ -16,49 +22,117 @@ trait HttpEventStreamHandle {
   def close(): Unit
 }
 
+class HttpEventStreamActorMetrics @Inject() (metrics: Metrics) {
+  val numberOfStreams: AtomicIntGauge =
+    metrics.gauge(metrics.name("api", getClass, "number-of-streams"), new AtomicIntGauge)
+}
+
 /**
   * This actor handles subscriptions from event stream handler.
   * It subscribes to the event stream and pushes all marathon events to all listener.
   * @param eventStream the marathon event stream
   */
-class HttpEventStreamActor(eventStream: EventStream, maxOutstandingMessages: Int) extends Actor {
+class HttpEventStreamActor(
+  eventStream: EventStream,
+  metrics: HttpEventStreamActorMetrics,
+  handleStreamProps: HttpEventStreamHandle => Props)
+    extends Actor {
   //map from handle to actor
-  private[http] var clients = Map.empty[HttpEventStreamHandle, ActorRef]
+  private[http] var streamHandleActors = Map.empty[HttpEventStreamHandle, ActorRef]
   private[this] val log = Logger.getLogger(getClass)
 
-  override def receive: Receive = {
-    case HttpEventStreamConnectionOpen(handle)   => addHandler(handle)
-    case HttpEventStreamConnectionClosed(handle) => removeHandler(handle)
-    case Terminated(actor)                       => removeActor(actor)
+  override def preStart(): Unit = {
+    metrics.numberOfStreams.setValue(0)
+    eventStream.subscribe(self, classOf[LeadershipStateEvent])
+  }
+
+  override def postStop(): Unit = {
+    eventStream.unsubscribe(self)
+    metrics.numberOfStreams.setValue(0)
+  }
+
+  override def receive: Receive = standby
+
+  // behaviours
+  private[this] val active: Receive = behaviour(acceptingNewConnections)
+  private[this] val standby: Receive = behaviour(rejectingNewConnections)
+
+  /**
+    * Helper method to create behaviours.
+    * The behaviours only differ in how they deal with new connections.
+    */
+  private[this] def behaviour(newConnectionBehaviour: Receive): Receive = {
+    // Prevent fatal warning about deriving type Any as type parameter
+    def orElse(receives: Receive*): Receive = receives.reduceLeft { _.orElse[Any, Unit](_) }
```

Comment:
Minor: For me the simpler version would be easier to understand: Seq(handleLeadership, cleanupHandlerActors, newConnectionBehaviour, warnAboutUnknownMessages).reduceLeft { _.orElse[Any, Unit](_) }